cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java

```diff
@@ -148,6 +150,147 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
+          scanner.close();
+          break;
+        }
+
+        offset += entityCodec.getEncodedSize(MetricsEntityType.CONTEXT);
+        String metricName = entityCodec.decode(MetricsEntityType.METRIC, rowKey, offset);
+
+        if (isContextQuery) {
+          // update the rowkey, by incrementing the next-level context part
+          ByteBuffer contextPart = ByteBuffer.wrap(rowKey);
+          contextPart.limit(entityCodec.getEncodedSize(MetricsEntityType.CONTEXT));
+          contextPart = getNextContextTarget(contextPart, query.getContextPrefix());
+          if (contextPart == null) {
+            scanner.close();
+            break;
+          }
+          //update the start-row to be row-key for next scan
+          startRow = rowKey;
+          metricsScanResults.add(contextStr);
+        } else {
+          // update the rowkey, by incrementing the metrics part
+          ByteBuffer metricsPart = ByteBuffer.wrap(rowKey);
+          metricsPart.position(entityCodec.getEncodedSize(MetricsEntityType.CONTEXT));
+          metricsPart.limit(entityCodec.getEncodedSize(MetricsEntityType.CONTEXT) +
+                              entityCodec.getEncodedSize(MetricsEntityType.METRIC));
+          metricsPart = stopKeyForPrefix(metricsPart);
+          if (metricsPart == null) {
+            scanner.close();
+            break;
+          }
+          //update the start-row to be row-key for next scan
+          startRow = rowKey;
+          metricsScanResults.add(metricName);
+        }
+      }
+      // no more matching rows found
+      scanner.close();
+    } while (rowResult != null);
+
+    return metricsScanResults;
+  }
+
+  /**
+   * Based on {@link co.cask.cdap.api.common.Bytes stopKeyForPrefix } but does the increment on the ByteBuffer instead.
+   * @param prefix
+   * @return
+   */
+  private static ByteBuffer stopKeyForPrefix(ByteBuffer prefix) {
+    for (int i = prefix.limit() - 1; i >= prefix.position(); i--) {
+      int unsigned = prefix.get(i) & 0xff;
+      if (unsigned < 0xff) {
+        byte result = prefix.get(i);
+        result += 1;
+        prefix.put(i, result);
+        return prefix;
+      }
+    }
+    // i.e. "read to the end"
+    return null;
+  }
+
+  /**
+   * Given the context and contextPrefix, this increments the context part after the contextPrefix, or the first context
+   * if no contextPrefix is provided. returns null if we cannot increment and have reached max value for the
+   * context part.
+   * @param context
+   * @param contextPrefix
+   * @return
+   */
+  private ByteBuffer getNextContextTarget(ByteBuffer context, String contextPrefix) {
+    int idSize = entityCodec.getIdSize();
+    // if null , we only increment the first part of context, otherwise we increment the part after the contextPrefix.
+    int targetContextLocation = contextPrefix == null ? 0 : StringUtils.countMatches(contextPrefix, ".") + 1;
```

Comment:
I don't think we use commons StringUtils anywhere.. Let's try not to bring another dependency.

Also, seems like you can calculate targetContextLocation once per query, no? Would make things more readable and efficient. Just create variable contextPartIndex or smth and use it inside while loop. Makes sense?